api/src/main/java/org/apache/iceberg/types/CheckCompatibility.java

```diff
@@ -250,6 +250,19 @@ public List<String> map(
     }
   }
 
+  @Override
+  public List<String> variant(Types.VariantType readVariant) {
+    if (currentType.isVariantType()) {
+      return NO_ERRORS;
+    }
+
+    // Currently promotion is not allowed to variant type
+    return ImmutableList.of(
+        String.format(
+            ": %s cannot be read as a %s",
+            currentType.typeId().toString().toLowerCase(Locale.ENGLISH), readVariant));
```

Comment:
Nit: `currentType` should work instead of this conversion. I know this was copied from primitive, but it isn't needed.

Reply:
Yeah. You are right. currentType already has toString() implementation.